plugins/webhooks/index.js

```diff
@@ -2,10 +2,76 @@
 
 const boom = require('boom');
 const joi = require('joi');
+const winston = require('winston');
 const workflowParser = require('screwdriver-workflow-parser');
 
 const WAIT_FOR_CHANGEDFILES = 1.8;
 
+/**
+ * Update admins array
+ * @param  {Object}    permissions  User permissions
+ * @param  {Pipeline}  pipeline     Pipeline object to update
+ * @param  {String}    username     Username of user
+ * @return {Promise}                Updates the pipeline admins and throws an error if not an admin
+ */
+function updateAdmins(permissions, pipeline, username) {
+    const newAdmins = pipeline.admins;
+
+    // Delete user from admin list if bad permissions
+    if (!permissions.push) {
+        delete newAdmins[username];
+        // This is needed to make admins dirty and update db
+        pipeline.admins = newAdmins;
+
+        return pipeline.update()
+            .then(() => {
+                throw boom.forbidden(`User ${username} `
+                + 'does not have push permission for this repo');
+            });
+    }
+
+    // Add user as admin if permissions good and does not already exist
+    if (!pipeline.admins[username]) {
+        newAdmins[username] = true;
+        // This is needed to make admins dirty and update db
+        pipeline.admins = newAdmins;
+
+        return pipeline.update();
+    }
+
+    return Promise.resolve();
+}
+
+/**
+ * Update admins array after check existance of pipeline, user and permission
+ * @param  {UserFactory}    userFactory     UserFactory object
+ * @param  {String}         username        Username of user
+ * @param  {String}         scmContext      Scm which pipeline's repository exists in
+ * @param  {Pipeline}       pipeline        Pipeline object
+ */
+async function checkAndUpdateAdmins(userFactory, username, scmContext, pipeline) {
+    try {
+        await userFactory.get({ username, scmContext })
```

Comment:
I think the above example is more simple when we use `await`.
```
try{
  const user = await userFactory.get({ username, scmContext });
  const userPermissions = await user.getPermissions(pipeline.scmUri);

  return updateAdmins(
                userPermissions,
                pipeline,
                username));

} catch (err) {...}
```

Reply:
It's nice idea. I fixed. Thank you.